File: backend/app/routers/documents.py
"""What documents can be drafted, and what each one needs.

The specs are the single source of truth for a document's fields, and three
things need them: the prompt, the validation of what the model returns, and the
client's cover page renderer. Serving them here means the client renders any
document generically without a second copy of the schemas.
"""

from fastapi import APIRouter, HTTPException, status

from .. import documents
from ..documents import DocumentSpec
from ..models import Camel
from ..templates import DRAFT_NOTICE, prose_for, read_template

router = APIRouter(prefix="/documents", tags=["documents"])


class DocumentSummary(Camel):
    """One entry in the catalogue, enough to list and choose from."""

    id: str
    name: str
    description: str


class DocumentDetail(Camel):
    """A document's spec, with the legal text it renders.

    Spec and template travel together because neither is useful alone, and
    fetching them separately would let a client render a spec against the wrong
    template.
    """

    spec: DocumentSpec
    #: Raw markdown of the Standard Terms, rendered verbatim.
    standard_terms: str
    #: The paragraph incorporating the Standard Terms by reference. Empty for
    #: documents with no cover page template of their own.
    preamble: str
    #: The CC BY 4.0 attribution, which the licence requires.
    attribution: str
    #: The warning that this is an unreviewed draft. Served rather than written
    #: into the client so the wording has one source.
    disclaimer: str = DRAFT_NOTICE


@router.get("")
def catalogue() -> list[DocumentSummary]:
    """Every document type, in the order the catalogue lists them."""
    return [
        DocumentSummary(id=spec.id, name=spec.name, description=spec.description)
        for spec in documents.SPECS
    ]


@router.get("/{document_id}")
def detail(document_id: str) -> DocumentDetail:
    """One document's spec and legal text."""
    spec = documents.get(document_id)
    if spec is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"There is no document type {document_id!r}.",
        )
    prose = prose_for(spec)
    return DocumentDetail(
        spec=spec,
        standard_terms=read_template(spec.template),
        preamble=prose.preamble,
        attribution=prose.attribution,
    )

File: frontend/lib/documents.ts
/**
 * The document types the app can draft, as the backend describes them.
 *
 * These mirror `backend/app/documents/schema.py`. The specs live on the server
 * because the prompt, the validation and this renderer all need them and one
 * copy is enough; the client fetches them rather than restating them.
 */

import { request } from './api'

export type FieldType = 'text' | 'longText' | 'date' | 'choice' | 'years'

/** One alternative a `choice` field can take. */
export interface ChoiceOption {
  id: string
  /** Cover page voice: a complete statement of what the parties agreed. */
  cover: string
  /** Standard Terms voice: a noun phrase that slots into a sentence. */
  reference: string
}

/** One value a document needs. */
export interface FieldSpec {
  id: string
  label: string | null
  hint: string | null
  type: FieldType
  required: boolean
  default: string
  /** Literal text printed before the value, e.g. "courts located in ". */
  prefix: string | null
  /** Printed instead of the value when an optional field is left blank. */
  emptyText: string | null
  guidance: string | null
  /** Labels this field supplies in the template's prose. */
  markers: string[]
  options: ChoiceOption[]
  /** For a choice whose wording includes `{years}`, the field holding it. */
  yearsField: string | null
}

/** A titled group of fields on the cover page. */
export interface SectionSpec {
  title: string
  hint: string | null
  fieldIds: string[]
}

export interface DocumentSpec {
  id: string
  name: string
  description: string
  template: string
  coverPageTemplate: string | null
  title: string
  attest: string
  parties: [string, string]
  shortName: string
  /** Every value the document captures, the signature block's included. */
  fields: FieldSpec[]
  sections: SectionSpec[]
  /** Terms the document negotiates that this app does not capture. */
  uncaptured: string[]
}

/** One entry in the catalogue, enough to list and choose from. */
export interface DocumentSummary {
  id: string
  name: string
  description: string
}

/** A spec together with the legal text it renders. */
export interface DocumentDetail {
  spec: DocumentSpec
  standardTerms: string
  preamble: string
  attribution: string
  /** The warning that this is an unreviewed draft, worded by the server. */
  disclaimer: string
}

export const fetchCatalogue = (): Promise<DocumentSummary[]> =>
  request('/documents')

export const fetchDocument = (id: string): Promise<DocumentDetail> =>
  request(`/documents/${id}`)

/** Looks a field up by id. */
export const fieldById = (
  spec: DocumentSpec,
  id: string,
): FieldSpec | undefined => spec.fields.find((field) => field.id === id)
